File: DataMiningSite/DataMiningAnalysis/src/main/scala/com/dfssi/dataplatform/analysis/config/EnvironmentConfig.scala
package com.dfssi.dataplatform.analysis.config

import org.apache.hadoop.fs.FileSystem
import org.apache.spark.Logging

import scala.xml.Elem

/**
  * Description:
  *
  * @author LiXiaoCong
  * @version 2018/6/21 10:15 
  */
class EnvironmentConfig(config: Map[String, Map[String, String]]) extends Serializable with Logging{

    require(config != null, "环境配置为null。")
    logInfo(s"环境配置细节如下：\n\t ${toString}")

    def getEnvironment(id: String): Map[String, String] = config.getOrElse(id, Map.empty[String, String])


    override def toString = s"EnvironmentConfig(${config})"
}

object EnvironmentConfig{

    val CONFIG_DIR = "/user/hdfs/config/env"
    val CONFIG_FILE = "environment.xml"

    def buildFromXml(el: Elem): EnvironmentConfig ={
        val configMap = (el \ "environment").map(node =>{
            val environment = node.asInstanceOf[Elem]
            val id = XmlReader.getAttr(environment, "id")
            require(id != null, s"标签${environment}中不存在id属性。")

            var paramMap: Map[String, String] = null
            val params = XmlReader.getNextSingleSubElem(environment, "params")
            if(params != null){
               paramMap = (params \ "param").map(n =>{
                    val param = n.asInstanceOf[Elem]
                    val name = XmlReader.getAttr(param, "name")
                    val value = XmlReader.getAttr(param, "value")
                    (name, value)
                }).toMap
            }

            (id, paramMap)
        }).toList.filter(_._2 != null).toMap

        new EnvironmentConfig(configMap)
    }

    def buildFromHdfsOrLocal(fs: FileSystem): EnvironmentConfig ={
        val elem = XmlReader.loadXmlConfig(fs, CONFIG_DIR, CONFIG_FILE)
        buildFromXml(elem)
    }

    def main(args: Array[String]): Unit = {
        val config = buildFromHdfsOrLocal(null)
        println(config)
    }

}

File: SchedulerSite/QuartzScheduler/src/main/java/com/dfssi/dataplatform/quartz/sync/EnvironmentConfig.java
package com.dfssi.dataplatform.quartz.sync;

import com.dfssi.common.WebHDFSUtil;
import com.google.common.collect.Maps;
import org.dom4j.Element;
import org.dom4j.io.SAXReader;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.StringReader;
import java.util.List;
import java.util.Map;

/**
 * Description:
 *
 * @author LiXiaoCong
 * @version 2018/6/22 16:21
 */
public class EnvironmentConfig {
    private final Logger logger = LoggerFactory.getLogger(EnvironmentConfig.class);

    private final String CONFIG_FILE = "/user/hdfs/config/env/environment.xml";

    private Map<String, Map<String, String>> environments = Maps.newHashMap();

    public EnvironmentConfig(String nameNodeHost, int nameNodeWebhdfsPort) throws Exception {

        String file = WebHDFSUtil.openfile(nameNodeHost, nameNodeWebhdfsPort, CONFIG_FILE);
        StringReader stringReader = new StringReader(file);

        Element rootElement = new SAXReader().read(stringReader).getRootElement();
        init(rootElement);

        logger.info(String.format("集群环境配置参数如下：\n\t %s", environments));
    }

    private void init(Element element){
        List<Element> envs = element.selectNodes("environment");
        List<Element> params;
        Map<String, String> configMap;
        for(Element env : envs){
            params = env.selectNodes("params/param");
            if(params != null){
                configMap = Maps.newHashMap();
                for(Element param : params){
                    configMap.put(param.attributeValue("name"),
                            param.attributeValue("value"));
                }
                environments.put(env.attributeValue("id"), configMap);
            }
        }
    }


    public Map<String, String> getEnvironment(String id){
        Map<String, String> environment = environments.get(id);
        if(environment == null) environment = Maps.newHashMap();
        return environment;
    }

    public static void main(String[] args) throws Exception {
        EnvironmentConfig environmentConfig = new EnvironmentConfig("devmaster", 50070);
    }

}
